scipy/constants/_constants.py

```diff
@@ -269,19 +268,21 @@ def convert_temperature(
     array([ 233.15,  313.15])
 
     """
+    xp = array_namespace(val)
+    val = xp.asarray(val)
     # Convert from `old_scale` to Kelvin
     if old_scale.lower() in ['celsius', 'c']:
-        tempo = np.asanyarray(val) + zero_Celsius
```

Comment:
Will use of `asarray` instead of `asanyarray` break anything that is supposed to be supported?

Reply:
According to https://stackoverflow.com/questions/59350359/any-examples-for-numpy-asanyarray-vs-asarray only if you are using subclasses of an `ndarray`, hopefully that is niche enough not to be an issue? This looks like a quirk of history rather than a change to in response to something https://github.com/scipy/scipy/commit/52f43e9f1b5391ddc7c30152318563c1f107a284

Reply:
Maybe worth trying to find the niche case - the alternative would be `val = np.asanyarray(val) if is_numpy(xp) else xp.asarray(val)` I think. `array_namespace` returns a `numpy` namespace for anything unrecognised but coercible by `np.asanyarray` (that isn't otherwise refused).

Dropping it should probably be an explicit deprecation otherwise.

Reply:
Or maybe we add a `subok` argument to `_asarray`?

Reply:
> Dropping it should probably be an explicit deprecation otherwise.

If it's advertised, sure, but I don't think it's needed if subclasses were supported by accident.

Reply:
It is advertised (as an `array_like`) - https://docs.scipy.org/doc/scipy/reference/generated/scipy.constants.convert_temperature.html#scipy.constants.convert_temperature

Reply:
Then according to the [NumPy glossary](https://numpy.org/doc/stable/glossary.html), `{1: 0}` and `scipy` are valid inputs, too.

![image](https://github.com/scipy/scipy/assets/6570539/78640ab7-b931-45db-930d-44010e239c91)

Reply:
🙃 well that's a much broader question of how strictly do we have to stick by poor documentation. I don't know the answer.

Reply:
Most of SciPy is advertised as `array like`, there are 2642 occurrences of `asarray` and 48 occurrences of `asanyarray` so I would be very surprised if anyone is relying on the behaviour of the slightly random selection of functions that use `asanyarray`. On the other hand, maybe I am being cavalier, happy to be convinced by other arguments!

Reply:
Right, in that case I agree that it isn't a big deal, but I think it would be worth it to officially 'fix' the documentation and deprecate the 48 occurrences of `asanyarray`. Thoughts?

Reply:
> On the other hand, maybe I am being cavalier, happy to be convinced by other arguments!

I don't think so. Starting this thread, I really just meant to check whether there were any tests or documentation that explicitly suggested that other arrays (e.g. masked arrays and matrices) should be accepted (as there are sometimes in stats). Since there aren't, I think it's safe enough.

Reply:
And yes, my point about `array_like` is that *technically* it's essentially meaningless. Pretty much the only thing that isn't allowed by `np.array` is a ragged array.
```python3
np.array([[1, 2, 3], [2, 3]])
```
And if we try to interpret it more strictly, it is reasonable to think that `np.array` *might actually by called* or that the object will be interpreted as a `np.ndarray` (not a subclass). So I think however you slice it, it's safe enough unless there's something more explicit in the documentation or tests.

Reply:
I do think it's a good idea to change the documentation, maybe as we go through these, to specify:
- `array` (if that's how array-API compatible arrays are to be identified in documentation, following the examples of the array API standard docs, e.g. [`abs`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.abs.html))
- dimensionality
- dtype

For example, `n-D real floating array`. But there would need to be a standard language for that, and I don't know that one exists.

Reply:
Yep, I've written somewhere about this before but not sure where. There are some orthogonal issues:
- the 48 occurrences of `np.asanyarray` should probably be replaced with `np.asarray` unless there really are documented special cases. This may or may not require deprecations, depending on how much the behaviour can be deemed undocumented.
- parameters are variably described as `ndarray` or `array_like`, but in both cases it usually means 'anything that `np.asarray` can make sense of'. This should be replaced with something like `array` as you suggest, along with the other information. 
- We need to keep supporting array-likes with `np` arrays, but they will not be supported for other backends. Hence, there is even more reason to get away from saying `array_like`. We can document elsewhere that array-likes will still be supported with NumPy as the assumed backend.

Reply:
Ndarray subclass support is not in great/consistent shape, but I also don't see the need to break it if a one-liner keeps things unchanged.

This is covered in gh-18286:

| Input type                | Return type          | When                          | Behavior                  | Notes                                                                               |
|---------------------------|----------------------|-------------------------------|---------------------------|-------------------------------------------------------------------------------------|
| numpy.ndarray subclasses  | same subclass type   | always                        |                           | so always use `np.asanyarray` checks                                                |

Reply:
How about following Jake's suggestion with `_asarray(x, ..., xp=None, ..., sub_ok=True)` which uses `np.asanyarray` instead of `np.asarray` for NumPy, but still `xp.asarray` for alternative backends?

Reply:
That sounds good to me.